### PR TITLE
Add SAS Token Authentication Support to Azure Repo Plugin

### PR DESCRIPTION
**Pull request by original-brownbear**
* Added setting for SAS token
* Added support for the token in tests
  * I have manually verified that the setting works with real Azure. I'm not 100% happy with the way I'm passing the SAS token setting to tests here, but I figured this might be the easiest approach for infra. The implemented approach allows to only test either a client with a key set or a client with set SAS token in a single Gradle run. This seemed less cumbersome than allowing to provide two separate sets of credentials and then adding code to Gradle to run the third party test goals twice. Also, for infra this is a reasonable approach I think because they can just copy the existing Jenkins job, remove the key setting and add an SAS token setting in its place in the copy => no effort required there outside of copying the Gradle target. 
* Relates #42117

### CODE REVIEW

File: plugins/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureStorageSettings.java
Comment (ywelsch) [inline plugins/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureStorageSettings.java]: can you add docs for the new setting as well?

File: plugins/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureStorageSettings.java
Comment (ywelsch) [inline plugins/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureStorageSettings.java]: add connectString instead here?

**Review by ywelsch — CHANGES_REQUESTED** [review]
I've left two smaller asks. In general, I think we should move away from the low-level connection string creation and use the objects provided by the SDK for that. We can do that as a follow-up, however, as that will possibly complicate backporting

File: plugins/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureStorageSettings.java
Comment (original-brownbear) [reply plugins/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureStorageSettings.java]: @ywelsch this was actually an on purpose removal, do we really want to add secure credentials in `toString` methods?

File: plugins/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureStorageSettings.java
Comment (original-brownbear) [reply plugins/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureStorageSettings.java]: Done in https://github.com/elastic/elasticsearch/pull/42982/commits/b7045d09631d23e8a2f4c8554d382f72768d8aac :)

**Comment by original-brownbear** [comment]
@ywelsch all addressed :)

**Review by ywelsch — APPROVED** [review]
LGTM. I would like a second pair of eyes on this (@tlrx or @andrershov) given the backport to 6.8.

File: docs/plugins/repository-azure.asciidoc
Comment (tlrx) [inline docs/plugins/repository-azure.asciidoc]: I think we should say a word about the type of SAS token and the required permissions

**Review by tlrx — CHANGES_REQUESTED** [review]
It looks great! I left some minor comments that need to be addressed before merging. I tested it both manually and using the 3rd party tests and it worked. 

There's an interesting point about adding multiple CI jobs for 3rd party tests vs. adding more gradle test tasks. I think it make sense here to add another CI job (and this is maybe what we'd like to do for S3 too).

File: docs/plugins/repository-azure.asciidoc
Comment (original-brownbear) [reply docs/plugins/repository-azure.asciidoc]: Ah my bad that was just a mistake in copy and pasting, thanks for spotting :)

File: plugins/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureStorageSettings.java
Comment (original-brownbear) [reply plugins/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureStorageSettings.java]: Unfortunately, for that we'd have to keep some field that holds this information. Maybe not worth it?

File: plugins/repository-azure/build.gradle
Comment (original-brownbear) [reply plugins/repository-azure/build.gradle]: Sure :)

File: plugins/repository-azure/qa/microsoft-azure-storage/build.gradle
Comment (original-brownbear) [reply plugins/repository-azure/qa/microsoft-azure-storage/build.gradle]: Right ... obviously in my local testing I just set the env vars to `""` when switching between auth methods :D and didn't notice.

File: docs/plugins/repository-azure.asciidoc
Comment (original-brownbear) [reply docs/plugins/repository-azure.asciidoc]: I think the type doesn't matter just the permissions (service as well as accounts SAS should work)? I added a comment on the permissions now :)

**Comment by original-brownbear** [comment]
Thanks @tlrx all points addressed I think/hope :)

File: plugins/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureStorageSettings.java
Comment (tlrx) [reply plugins/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureStorageSettings.java]: +1

File: docs/plugins/repository-azure.asciidoc
Comment (tlrx) [reply docs/plugins/repository-azure.asciidoc]: Talked to Armin using another channel. I'd expect to see some `create` or `list` permissions too here

Comment (original-brownbear) [reply docs/plugins/repository-azure.asciidoc]: Alright, updated this with all the details now. `create` we don't need since it's implied by `write` for the operations we execute https://docs.microsoft.com/en-us/rest/api/storageservices/Constructing-an-Account-SAS?redirectedfrom=MSDN#blob-service (as a matter of fact it is for all operations on the blob service). But we needed `list`.
Also, I added some detail on the service and resource type the token has to apply to.

**Review by tlrx — APPROVED** [review]
LGTM, thanks!

**Comment by original-brownbear** [comment]
thanks @tlrx !

**Comment by original-brownbear** [comment]
Just FYI, I raised an infra issue for getting the new CI job going.

**Pull request merged**
